fastestimator/trace/io/model_eval.py

```diff
@@ -0,0 +1,451 @@
+            equal to False.
+    """
+    def __init__(self, description: str, criteria: Callable, aggregate: bool = True, fail_threshold: int = 0) -> None:
+        self.description = description
+        self.criteria = criteria
+        self.criteria_inputs = inspect.signature(criteria).parameters.keys()
+        self.aggregate = aggregate
+        if not self.aggregate:
+            self.fail_threshold = fail_threshold
+
+    def clean_result(self) -> None:
+        if not self.aggregate:
+            self.result = []
+            self.fail_id = []
+
+
+@traceable()
+class ModelEval(Trace):
+    """Automate testing and report generation.
+
+    This trace will examine all input TestCases in the test mode and generate a PDF report and a JSON test result.
+
+
+    Args:
+        test_cases: TestCase object or list of TestCase objects.
+        save_path: Where to save the output directory.
+        test_title: Title of the test.
+        data_id: Data instance ID key. If provided, then per-instance test will return failure instance ID.
+    """
+    def __init__(self,
+                 test_cases: Union[TestCase, List[TestCase]],
+                 save_path: str,
+                 test_title: str = "Test",
+                 data_id: str = None) -> None:
+
+        self.test_title = test_title
+        self.test_cases = to_list(test_cases)
+        self.instance_cases = []
+        self.aggregate_cases = []
+        self.data_id = data_id
+
+        all_inputs = set()
+        for case in self.test_cases:
+            all_inputs.update(case.criteria_inputs)
+            case.clean_result()
+            if case.aggregate:
+                self.aggregate_cases.append(case)
+            else:
+                self.instance_cases.append(case)
+
+        if self.data_id:
+            all_inputs.update([data_id])
+
+        path = os.path.normpath(save_path)
+        path = os.path.abspath(path)
+        root_dir = os.path.dirname(path)
+        report = os.path.basename(path) or 'report'
+        report = report.split('.')[0]
+        self.save_dir = os.path.join(root_dir, report)
+        self.src_dir = os.path.join(self.save_dir, "resources")
+        os.makedirs(self.save_dir, exist_ok=True)
+        os.makedirs(self.src_dir, exist_ok=True)
+
+        super().__init__(inputs=all_inputs, mode="test")
+
+    def on_begin(self, data: Data) -> None:
+        self._sanitize_report_name()
+        self._initialize_json_summary()
+
+    def on_batch_end(self, data: Data) -> None:
+        for case in self.instance_cases:
+            result = case.criteria(*[data[var_name] for var_name in case.criteria_inputs])
+            if not isinstance(result, np.ndarray):
+                raise TypeError(f"In test with description \'{case.description}\': "
+                                "Criteria return of per-instance test needs to be ndarray with dtype bool.")
```

Comment:
this is different from the doc string for TestCase

Reply:
i am going to change the docstring to:

criteria: A function to perform the test. In aggregate tests, the function needs to return True when the test
            passes and False when it fails. In per-instance tests, the function needs to returns a ndarray of bool, where
            entries show corresponding test results. (True if the test of that data instance passes; False if the test fails).